CIP-0095/README.md

```diff
@@ -0,0 +1,771 @@
+registered DReps.
+
+The methods described here should not be considered the only definitive method
+of generating DRep credentials. Rather these methods should be employed to
+derive keys for non-script registered DReps.
+
+#### Derivation
+
+Here we describe DRep Key derivation as it attains to Cardano wallets who follow
+the
+[CIP-1852 | HD (Hierarchy for Deterministic) Wallets for Cardano](https://github.com/cardano-foundation/CIPs/blob/master/CIP-1852/README.md)
+standard.
+
+To differentiate DRep credentials from other Cardano keys the derivation path
+must follow:
+
+`m / 1718' / 1815' / account' / chain / address_index`
+
+> **Note** `1718` was the year that François-Marie adopted the pseudonym
+> Voltaire.
+
+We strongly suggest that a maximum of one set of DRep credentials should be
+associated with one wallet account, this can be achieved by setting `chain=0`
+and `address_index=0`. Thus avoiding the need for DRep Key discovery.
+
+We believe the overhead that would be introduced by "multi-DRep" accounts is an
+unjustified expense. Future iterations of this specification may expand on this,
+but at present this is seen as unnecessary.
+
+#### Tooling
+
+Supporting tooling should clearly label these key pairs as "CIP-95 DRep Keys".
+
+Bech32 prefixes of `drep_sk` and `drep_vk` should be used, as described in
+[CIP-0005](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0005/README.md).
+
+Examples of acceptable `keyType`s for supporting tools:
+
+| `keyType`                          | Description           |
+| ---------------------------------- | --------------------- |
+| `CIP95DRepSigningKey_ed25519`      | DRep Signing Key      |
+| `CIP95DRepVerificationKey_ed25519` | DRep Verification Key |
+
+For hardware implementations:
+
+| `keyType`                          | Description                    |
+| ---------------------------------- | ------------------------------ |
+| `CIP95DRepVerificationKey_ed25519` | Hardware DRep Verification Key |
+| `CIP95DRepHWSigningFile_ed25519`   | Hardware DRep Signing File     |
+
+### Data Types
+
+From
+[CIP-30's Data Types](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#data-types)
+we only inherit:
+
+- [cbor\<T>](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#cbort)
+
+#### PubDRepKey
+
+```ts
+type PubDRepKey = string;
+```
+
+A hex-encoded string representing 32 byte Ed25519 DRep public key, as described
+in [DRep Key](#DRep-key).
+
+#### PubStakeKey
+
+```ts
+type PubStakeKey = string;
+```
+
+A hex-encoded string representing 32 byte Ed25519 public key used as a staking
+credential.
+
+#### UnsignedTransaction
+
+```ts
+type UnsignedTransaction = string;
+```
+
+A hex-encoded string representing a CBOR transaction that is completely unsigned
+(has an empty transaction witness set).
+
+This data model is used to represent transactions which contain, for example:
+DRep registration certificates and the client wishes the wallet inspect, add
```

Comment:
are these certificates supposed to be included in the transaction as metadata or will there be a dedicated field in the Cardano transaction body?

If the former, rather than the "UnsignedTransaction" I'd expect there to be an "Unsigned metadata" property representing the unsinged vote delegation, drep registration etc. metadata.

If the latter (i.e. the governance actions are part of the Cardano tx body), I'd expect the CIP-30 `submitTransaction()` call to already cover the use case as it already should be able to work with any raw transaction body (as long as the underlying wallet has the keys to properly sign it)

Are there already any specs for the respective transactions (vote delegation, drep registration, ...)? I couldn't find them in this CIP but maybe I just missed that

Reply:
See draft [Conway era ledger spec](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/conway/test-suite/cddl-files/conway.cddl)

Governance actions and votes are added in new transaction fields. Vote delegation, DRep registration and DRep retirement are new certificates.

Reply:
I see - so far when new fields were added to txs, wallets started parsing them as any DApp can start using them and it's in the interest of wallets to work with as many DAps as possible and having a call to sign/submit transactions within CIP-95 that is basically just a limited version of the CIP-30 call seems redundant to me and would actually be additional work for us as wallet developers as apart from the parsing alone to show the tx details, we'd also need to implement additional validation to reject non-compliant txs. So rather than adding a new "sign/submit transaction" method, I'd suggest specifying in this CIP that wallets supporting the CIP-95 extension need to be able to sign/parse conway-era transactions

When it comes to certificates, I'd consider generalizing the individual registration calls to just "signCertificateTx()" which could take one (or more, if it makes sense) certificate(s), parse them (as wallets already parse transactions) and produce a valid transaction containing the signed certificate(s), which should also give room for easier introduction of new kinds of certificates in the future without increasing the API surface, just the extension number would be increased to signal support of certain txs/certificates, or a "specVersion" field introduced to the CIP-95 extension API, to signal to the DApp what certificates/txs the wallet is supposed to support. What do you think?

Reply:
the latter ("signCertificateTx") may be a bit tricky as it may not clear how to represent the unsigned certificate in all cases, but extending the signTransaction call in CIP-30 should be IMHO a reasonable expectation from the wallets which should actually be less effort than the proposed dedicated calls

Reply:
Okay, so instead CIP95 could do its own `signTx` which replaces CIP30's `.signTx` to provide support for Conway Tx/certs? Would you want to support all types of certificate that are being introduced in Conway ledger era OR just the ones needed for Ada Holders/DReps as I have focussed on?

Reply:
I just checked and the certificates are indeed part of the transaction body for the conway era: https://github.com/input-output-hk/cardano-ledger/blob/master/eras/conway/test-suite/cddl-files/conway.cddl#L260 and I understand the CIP-30 signTransaction call to be era-agnostic, it just takes the raw tx and does whatever is needed to sign/show it.

In the past, as the Alonzo and Babbage eras were launched, the wallets' CIP-30 signTransaction calls evolved naturally with them (naturally incetivized by the DApps wanting to leverage the new functionalities) and I don't see why wouldn't that be the case with Conway, too, especially for any wallets that would commit to support the CIP-95 extension, so IMHO it's enough to specify what transaction fields/types the wallet must handle in its CIP-30 signTransaction call to be "CIP-95" compliant without resorting to specifying a whole new "cip95" signTx call which would effectively be just an alias for the CIP-30 one, if I'm not missing anything

Reply:
My main concern with “evolved naturally” is the lack of versioning resulting in a mismatch between expected wallet capabilities and actual capabilities. Using CIP-95 as a Conway version could make sense?

Reply:
I'd consider a valid way of "versioning" the wallet's API the exposure of the "CIP-95" extension itself which would signal the readiness of the CIP-30 signTransaction call to the respective wallet. Can you clarify what additional benefit would the signTransaction call within the CIP-95 extension have?

My concern is that by haveing a separate "signTx" call in the CIP-95 extension, we'd end up with two different calls which would do the same thing which would only result just in ambiguous implementations of the signing of transactions by DApps (some would go with the CIP-30 call, e.g. due to its use in the past, some would use the "CIP-95" call) and wallet developers may also be confused if the CIP-30 call should be any different from the CIP-95 one, which would be counterproductive overall.

I'd suggest having in this CIP specs of the exact transaction types, ideally even with examples, that wallets need to support to consider themselves "CIP-95" compliant (which the wallet would signal to the DApp with the CIP-95 extension flag being exposed), IMHO that would add most value both to DApp and wallet developers and remove the ambiguity that we might worry about regarding the lack of standardized versioning of the CIP-30 API itself

Reply:
> Can you clarify what additional benefit would the signTransaction call within the CIP-95 extension have?

For me, beyond the simplicity, this prevents the need for other proposals to support the Conway artefacts which this proposal lacked, for example: Constitutional Committee member hot/cold key registration certificates.

> I'd suggest having in this CIP specs of the exact transaction types, ideally even with examples, that wallets need to support to consider themselves "CIP-95" compliant (which the wallet would signal to the DApp with the CIP-95 extension flag being exposed), IMHO that would add most value both to DApp and wallet developers and remove the ambiguity that we otherwise might worry about regarding the lack of standardized versioning of the CIP-30 API itself

This is great suggestion, I fully agree that this, I will likely create a test vectors file with such details similar to the [test vectors included in CIP-36](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0036/test-vector.md).